econml/dml_iv.py

```diff
@@ -0,0 +1,760 @@
+                         _FirstStageWrapper(model_T_X, discrete_treatment),
+                         _FirstStageWrapper(model_T_XZ, discrete_treatment),
+                         _FinalWrapper(model_final,
+                                       fit_cate_intercept=fit_cate_intercept,
+                                       featurizer=featurizer,
+                                       use_weight_trick=False),
+                         n_splits=n_splits,
+                         discrete_instrument=discrete_instrument,
+                         discrete_treatment=discrete_treatment,
+                         random_state=random_state)
+
+
+class NonParamDMLIV(_BaseDMLIV):
+    """
+    A child of the _BaseDMLIV class that allows for an arbitrary square loss based ML
+    method in the final stage of the DMLIV algorithm. The method has to support
+    sample weights and the fit method has to take as input sample_weights (e.g. random forests), i.e.
+    fit(X, y, sample_weight=None)
+    It achieves this by re-writing the final stage square loss of the DMLIV algorithm as:
+        \sum_i (E[T|X_i, Z_i] - E[T|X_i])^2 * ((Y_i - E[Y|X_i])/(E[T|X_i, Z_i] - E[T|X_i]) - theta(X))^2
+    Then this can be viewed as a weighted square loss regression, where the target label is
+        \tilde{Y}_i = (Y_i - E[Y|X_i])/(E[T|X_i, Z_i] - E[T|X_i])
+    and each sample has a weight of
+        V(X_i) = (E[T|X_i, Z_i] - E[T|X_i])^2
+    Thus we can call any regression model with inputs:
+        fit(X, \tilde{Y}_i, sample_weight=V(X_i))
+    """
+
+    def __init__(self, model_Y_X, model_T_X, model_T_XZ, model_final,
+                 featurizer=None, fit_cate_intercept=True,
+                 n_splits=2, discrete_instrument=False, discrete_treatment=False):
+        """
+        Parameters
+        ----------
+        model_Y_X : model to predict E[Y | X]
+        model_T_X : model to predict E[T | X]
+        model_T_XZ : model to predict E[T | X, Z]
+        model_final : final model for predicting \tilde{Y} from X with sample weights V(X)
+        n_splits : number of splits to use in cross-fitting
+        discrete_instrument : whether to stratify cross-fitting splits by instrument
+        discrete_treatment : whether to stratify cross-fitting splits by treatment
+        """
+        super().__init__(_FirstStageWrapper(model_Y_X, False),
+                         _FirstStageWrapper(model_T_X, discrete_treatment),
+                         _FirstStageWrapper(model_T_XZ, discrete_treatment),
+                         _FinalWrapper(model_final, fit_cate_intercept, featurizer, True),
+                         n_splits=n_splits,
+                         discrete_instrument=discrete_instrument,
+                         discrete_treatment=discrete_treatment)
+
+
+class _BaseDRIV(_OrthoLearner):
+
+    """
+    The _BaseDRIV algorithm for estimating CATE with IVs. It is the parent of the
+    two public classes {DRIV, ProjectedDRIV}
```

Comment:
I assume `DRIV, ProjectedDRIV, IntentToTreatIV` will be part of another PR?

Reply:
Yes.